README.md
# SSBpris

This package contains functions for calculating price indices and their variances. It also contains useful function for allocating samples used for price indices. 

More on allocation of samples for price indices can be found here (norsk internal SSB document): https://wiki.ssb.no/display/s880/Standard+utvalgsplanlegging+i+Norsamu?preview=/32669820/32964617/SBS-utvalgsplanlegging.pdf 


NEWS.md
# 0.1.2

* New licence

tests/testthat.R
library(testthat)
library(SSBpris)

test_check("SSBpris")


tests/testthat/test-calc.R
# Testing of price index calculations
#source("R/Calculation_funcs.R")

test_that("CalcInd returns correct value", {
  data(priceData)
  ind <- CalcInd(data = priceData, baseVar = "b1", pVar = "p1", groupVar = "varenr", wVar = "weight", 
                 consumVar = "coicop", type = "dutot")
  expect_equal(as.numeric(ind[1]), 30.06709, tolerance = 1E-4)
})


tests/testthat/test-alloc.R
test_that("Allocation using FillStrataInd works", {
  
  data(priceData)
  
  s2 <- CalcIndS2(data = priceData, baseVar = "b1", pVar = "p1", groupVar = "varenr", 
                  type = "carli")$s2
  # Calculate bhg matrix
  ngh <- as.matrix(table(priceData$nace3, priceData$varenr))
  mh <- matrix(rep(table(priceData$nace3), 200), 5, 200)
  bhg <- ngh/mh 
  
  # Specify variance function for n (number of observations) - unweighted
  VarP <- function(n){ return(sum(s2 / n , na.rm = TRUE)) }
  
  # Run algorithm filling up one at a time for sample size of 300
  a <- FillStrataInd(totm = 300, bhg, totVar = VarP, min_n = 1, steps = 1, 
                     direction = "up")
  
  expect_equal(as.numeric(a$alloc[1]), 59)
})




R/testmidl.R
library(AllocSN)

data <- priceData
data <- testdata
baseVar <- "b1"
pVar <- "p1"
type <- "jevons"
groupVar <- "varenr_fac"
consumVar <- "coicop"
wVar = "weight"



testdata <- priceData

testdata$varenr_fac <- factor(testdata$varenr)
levels((testdata$varenr_fac))

testdata <- testdata[(testdata$varenr)<=100,]
levels((testdata$varenr_fac))

testdata <- testdata[order(testdata$dufNr),]

testdata$agg1 <- ifelse(testdata$coicop==1, "En",
                    ifelse(testdata$coicop==2, "To",
                           ifelse(testdata$coicop==3, "Tre",
                                  ifelse(testdata$coicop==4, "Fire",
                                         ifelse(testdata$coicop==5, "Fem",
                                                ifelse(testdata$coicop==6, "Seks",
                                                       ifelse(testdata$coicop==7, "Sju",
                                                              ifelse(testdata$coicop==8, "Åtte",
                                                                     ifelse(testdata$coicop==9, "Ni",
                                                                            ifelse(testdata$coicop==10, "Ti", 11))))))))))





testdata[2,"varenr_fac"] <- ""
testdata[1,"varenr"] <- NA
testdata[3,"b1"] <- NA
testdata[4,"p1"] <- NA
testdata["2","coicop"] <- 1
testdata[,"coicop"] <- NA
testdata[6,"weight"] <- NA
testdata[7,"weight"] <- 0
testdata[testdata$coicop==1,"weight"] <- 0

x <- CalcInd(data=testdata, baseVar="b1", pVar="p1", type="jevons", groupVar="varenr", consumVar = "coicop", 
                     wVar = "weight")

x <- CalcInd(data=juks, baseVar="b1", pVar="p1", type="jevons", groupVar="varenr", consumVar = "coicop", 
             wVar = "weight")

x <- CalcInd(data=testdata, baseVar="b1", pVar="p1", type="jevons", groupVar="varenr", consumVar = "coicop", 
        )




R/Calculation_funcs.R

#' Jevons index caluclation
#'
#' @param x Vector of base prices.
#' @param y Vector of prices in the statistical time period. 
#'
#' @return Price index (L-type)
#' @export
Jevons <- function(x, y){
  z <- y / x
  zz <- exp(mean(log(z[is.finite(log(z))]), na.rm=T)) #Tian Yi Stack overflow - ta ut 0 og NA verdi
  return(zz)
}

#' Dutot index calculation
#'
#' @param x Vector of base prices.
#' @param y Vector of prices in the statistical time period.
#' @param na.rm Logical on whether to remove missing values for x or y in the calculation.
#' @param ze.rm Logical on whether to remove missing values for the instrument variable for the calculations.
#'
#' @return Price index (L-type)
#' @export
Dutot <- function(x, y, na.rm = TRUE, ze.rm = TRUE){
  if (ze.rm) {
    zero <- (x != 0)
    x <- x[zero]
    y <- y[zero]
  }
  if (na.rm){
    naVect <- (!is.na(x) & !is.na(y))
    x <- x[naVect]
    y <- y[naVect]
  }
  return((sum(y, na.rm = T) / sum(x, na.rm = T)))
}

#' Carli index calculation
#'
#' @param x Vector of base prices.
#' @param y Vector of prices in the statistical time period. 
#' @param na.rm Logical on whether to remove missing values for x or y in the calculation.
#' @param ze.rm Logical on whether to remove missing values for the instrument variable for the calculations. 
#'
#' @return Price index (L-type)
#' @export
Carli <- function(x, y, na.rm = TRUE, ze.rm = TRUE){
  if (ze.rm) {
    zero <- (x != 0)
    x <- x[zero]
    y <- y[zero]
  }
  if (na.rm){
    naVect <- (!is.na(x) & !is.na(y))
    x <- x[naVect]
    y <- y[naVect]
  }
  return(sum(y / x) / length(y))
}


#' Calculation of the estimate for a price index
#' 
#' Calculation of a price index
#'
#' @param data The dataset
#' @param baseVar The variable name for the base goods price.
#' @param pVar The variable name for the goods price.
#' @param type The type of index to calculate. Choose from: 'carli', 'dutot', 'jevons'.
#' @param groupVar The variable used for grouping observations - sometimes called the elementary group
#' @param consumVar The variable used for grouping elementary groups. This may be the publishing level.
#' @param wVar The variable used to weight the elementary groups up to the consumer publishing level. 
#'
#' @return   The estimates for the index are returned as a vector of length equal to the number of groups (in groupVar or consumVar if specified).
#' @export
#'
#' @examples{
#' data(priceData)
#' # Calculate index for consumer groups
#' CalcInd(data = priceData, baseVar = "b1", pVar = "p1", groupVar = "varenr", wVar = "weight", 
#'   consumVar = "coicop", type = "dutot")
#' CalcInd(data = priceData, baseVar = "b1", pVar = "p1", groupVar = "varenr", wVar = "weight", 
#'   consumVar = "coicop", type = "carli")
#' CalcInd(data = priceData, baseVar = "b1", pVar = "p1", groupVar = "varenr", wVar = "weight", 
#'   consumVar = "coicop", type = "jevons")
#'   
#' # Calculate index for elemenatry groups (weighted)
#' CalcInd(data = priceData, baseVar = "b1", pVar = "p1", groupVar = "varenr", wVar = "weight", 
#'    type = "dutot")
#'    
#'    }
CalcInd <- function (data, baseVar, pVar, type, groupVar, consumVar = NULL, 
                     wVar = NULL) 
{
  # change to factor
  if (!is.factor(data[, groupVar])) {
    data[, groupVar] <- factor(data[, groupVar])
  }
  
  if (any(is.na(data[, groupVar]))) 
    warning("groupVar has missing values. These observations will be removed.", call. = FALSE)
  data <- data[!is.na(data[, groupVar]),]
  
  
  if (is.null(consumVar) | missing(consumVar)) {
    warning("No consumer group variable was specified so an index was calculated for each elementary group.", 
            call. = FALSE)
  } else {
    if (any(is.na(data[, consumVar]))) 
    warning("consumVar has missing values. These observations will be removed.", call. = FALSE)
    data <- data[!is.na(data[, consumVar]),]
    }
  
  
  if (is.null(wVar)) {
    warning("No weight variable was specified so all elements have been given a weight = 1", 
            call. = FALSE)
    data[, wVar] <- rep(1, nrow(data))
  } else {
    if (any(!is.finite(data[, wVar]))) #is.finite is FALSE for NA, NaN, Inf, -Inf 
      warning("wVar has missing or invalid values. These observations will be removed.", call. = FALSE)
    data <- data[is.finite(data[, wVar]),]
    
  }
  
  if (is.null(data)) {
    stop("Dataset is empty")
  }else{ 
    if (nrow(data)==0)
    stop("There are no usable observations in the dataset")
  }
  
  data <- droplevels(data) #removing unused levels to ensure ind and tab below have same length
  ww <- data[, wVar]
  
  # Check all weights are equal within groups
  wi <- ave(ww, data[, groupVar], FUN = mean) #wi has same length as ww
  if (!all(wi == ww)){
    warning("Not all weights were equal within groups!! A mean weight is being used")
  }
  
  
  # Check weights add to 1
  wg <- tapply(wi, data[, groupVar], FUN=mean)
  
  # wg has length equal to the number of levels in groupVar 
  # (all levels are actually occuring in the dataset because unused levels have been dropped)
  # wg is ordered by groupVar
  # wg has names equal to the levels of groupVar
  
  #Scaling wg
  while (!isTRUE(all.equal(sum(wg), 1))){
    warning("Elementary group weights did not add to one and have been scaled.")
    wg <- wg/ sum(wg)
  }
  
  groups <- levels(data[, groupVar])
  # groups is ordered by groupVar
  
  #ind <- array(NA, length(groups), 1)
  ind <- rep(NA, length(groups))
  for (i in 1:length(groups)) {
    d <- data[, groupVar] == groups[i]
    if (type == "jevons") {
      Pi <- Jevons(data[d, baseVar], data[d, pVar])}
    if (type == "carli") {
      Pi <- Carli(data[d, baseVar], data[d, pVar])}
    if (type == "dutot") {
      Pi <- Dutot(data[d, baseVar], data[d, pVar])}
    
    ind[i] <- Pi * wg[groups[i]] #Pi is multiplied by corresponding weight - vil vi ha det?
    #ind[i] <- Pi
  }
  names(ind) <- groups
  
  if (!is.null(consumVar) | !missing (consumVar)) {
    tab <- unique(data.frame(data[, groupVar], data[, consumVar])) #only the levels actually present in data will be in tab
    tab <- tab[order(tab[, 1]), ]
    
    if (any(duplicated(tab[, 1]))) 
      stop("Some elementary groups contain elements that are found in several consumer groups. These should be exclusive.")
   
    #If no level of groupVar is duplicated, the number of rows in tab equals the number of levels in groupVar
    #so ind and tab[,2] have the same length
    
    #Could still have missing or invalid prices (NA, NaN, Inf, -Inf), or sumwg=0
    
    sumind <- tapply(ind, tab[, 2], sum)
    #sumind <- tapply(wg*ind, tab[, 2], sum)
    sumwg <- tapply(wg, tab[, 2], sum) #Sum weights in each consumer group
    ind <- sumind/sumwg #Dividing by sumwg is the same as scaling weights to one
    
  }
  
    return(ind)
}


#' Calculation of variance/sigma squared for price index
#' 
#' Calculation of sigma squared for a price index.
#' 
#' 
#' @param data The dataset
#' @param baseVar The variable name for the base goods price.
#' @param pVar The variable name for the goods price.
#' @param groupVar The variable used for grouping observations - sometimes called the elementary group
#' @param type The type of index to calculate. Choose from: 'carli', 'dutot', 'jevons'
#' @return 
#'   \item{s2}{sigma squred values for each strata group}
#'   \item{ai}{adjustment factor used in calculations} 
#'   \item{zmat}{Matrix of z values which can be used to check for outliers}
#'   
#' @export
#'   
#' @examples {
#' # Call in test dataset for an index
#' data(priceData)
#' 
#' # Calculate s2 for an index
#' ss <- CalcIndS2(data = priceData, baseVar = "b1", pVar = "p1", groupVar = "varenr", 
#' type = "carli")
#' ss <- CalcIndS2(data = priceData, baseVar = "b1", pVar = "p1", groupVar = "coicop", 
#' type = "jevons")
#' 
#' # print out s2 values
#' ss$s2
#' }
#' 
CalcIndS2 <- function (data, baseVar, pVar, groupVar, 
                       type = c("carli", "dutot", "jevons")) {
  if (type == "Carli") type <- "carli"
  if (type == "Dutot") type <- "dutot"
  if (type == "Jevons") type <- "jevons"
  if (is.null(type) | !type %in% c("carli", "dutot", "jevons")) 
    stop("you need to specify a type which is 'carli', 'dutot' or 'jevons'")
  x <- data[, baseVar]
  y <- data[, pVar]
  gg <- data[, groupVar]
  if (!is.factor(gg)) gg <- factor(gg)
  group <- levels(gg)
  s2 <- rep(NA, length(group))
  ai <- rep(NA, length(group))
  zmat <- NULL
  for (i in 1:length(group)) {
    d <- (gg == group[i])
    if (type == "carli") {
      Pi <- Carli(x[d], y[d])
      zij <- y[d]/x[d] - Pi
      ai[i] <- 1
    }
    if (type == "dutot") {
      Pi <- Dutot(x[d], y[d])
      zij <- y[d] - x[d] * Pi
      ai[i] <- mean(x[d[!is.na(x) & x != 0]])
    }
    if (type == "jevons") {
      Pi <- Jevons(x[d], y[d])
      zij <- log(y[d]/x[d]) - log(Pi)
      ai[i] <- 1/Pi^2
    }
    s2[i] <- stats::var(zij)
    if (length(zij) > 0) zmat <- rbind(zmat, data.frame(zij, group[i]))
  }
  colnames(zmat)[2] <- "group"
  names(s2) <- group
  names(ai) <- group
  return(list(s2 = s2, ai = ai, zmat = zmat))
}


R/data_documentation.R

#' Price test dataset
#' 
#' This dataset is a test dataset containing a small sample of 1000 goods
#' and prices. Variables include goods price over 3 periods and the base price
#' for the three equivalent periods.
#'
#' @usage data(priceData)
#' @format Dataset containing the variables::
#' \describe{
#'   \item{dufNr}{company number}
#'   \item{nace5}{industry group (5 digits)} 
#'   \item{nace3}{industry group (3digits)} 
#'   \item{varenr}{identity number for the specific goods}
#'   \item{coicop}{grouping variable for goods which are similar. This is the publishing level.} 
#'   \item{weight}{Variable specifying the weight that should be given based on the type of goods.} 
#'   \item{p1}{price in the first period}
#'   \item{p2}{price in the second period} 
#'   \item{p3}{price in the third period}
#'   \item{b1}{base price which should be used in the first period}
#'   \item{b2}{base price which should be used in the second period}
#'   \item{b3}{base price which should be used in the third period}
#' }
"priceData"


R/Allocation_funcs.R

#' Optimal allocation of sample sizes by filling up or down for an index
#' 
#' This function finds an optimal allocation by filling up or down one (or
#' several) units at a time in a stratum(a).
#' 
#' 
#' @param totm The total desired sample size as the number of units to select 
#' (NOT the number of desired observations).
#' @param bhg A matrix giving the average number of observations in each elementary 
#' group in each stratum. See Zhang 2015 for more details.
#' @param totVar The variance to the index as a function of n (number of observations).
#' @param min_n A minimum sample size in each stratum. This can be a single number 
#' used in all strata or a vector of numbers (one for each stratum).
#' @param max_n A maximum sample size in each stratum. This can be a single number 
#' used in all strata or a vector of numbers (one for each stratum).
#' @param steps The number of strata to fill up at each iteration. Default is set to 1.
#' @param rand specify the number of strata to consider for random selection fill up. 
#' For example rand = 10 will randomly select 1 (or the number specified in steps) strata 
#' to fill up based on the top 10 using probabilities proportional to the difference 
#' between the variances. If missing, the algorithm will chose the top strata. Currently 
#' not programmed for direction = 'down'
#' @param direction Specify the direction the algorithm should work: "up" to
#' fill up from 0 or a minimum value or "down" to remove one unit at a time
#' from a maximum starting point.
#' @return 
#' \item{alloc}{A vector with the new sample sizes in each strata. }
#' \item{totvar}{Total variance}
#' 
#' @export
#' @examples{
#'  
#' # call in test dataset
#' data(priceData)
#' 
#' # Calculate index for consumer groups
#' CalcInd(data = priceData, baseVar = "b1", pVar = "p1", groupVar = "varenr", wVar = "weight", 
#'         consumVar = "coicop", type = "dutot")
#' 
#' # Calculate s2 for index
#' s2 <- CalcIndS2(data = priceData, baseVar = "b1", pVar = "p1", groupVar = "varenr", 
#'                 type = "carli")$s2
#' 
#' # Calculate bhg matrix
#' ngh <- as.matrix(table(priceData$nace3, priceData$varenr))
#' mh <- matrix(rep(table(priceData$nace3), 200), 5, 200)
#' bhg <- ngh/mh 
#' 
#' # Specify variance function for n (number of observations) - unweighted
#' VarP <- function(n){ return(sum(s2 / n , na.rm = TRUE)) }
#' 
#' # Run algorithm filling up one at a time for sample size of 300
#' a <- FillStrataInd(totm = 300, bhg, totVar = VarP, min_n = 1, steps = 1, 
#'  direction = "up")
#' a
#' 
#' # Run algorithm filling down one at a time
#' max_n <- table(priceData$nace3)
#' a <- FillStrataInd(totm = 300, bhg = bhg, totVar = VarP, min_n = 1, max_n = max_n, 
#'  steps = 1, direction = "down")
#' a
#' }
FillStrataInd <- function(totm, bhg, totVar, min_n, max_n, steps = 1, rand, direction = "up"){
  if (missing(min_n)) min_n <- rep(1, nrow(bhg))
  if (length(min_n) == 1) min_n <- rep(min_n, nrow(bhg))
  if (missing(max_n) & direction == "down") stop("max_n must be specified for down direction")
  if (direction == "down" & !missing(rand)) stop("rand option currently not available for down direction")
  if (missing(max_n)) max_n <- rep(Inf, nrow(bhg))
  if (!is.vector(max_n)) max_n <- as.vector(max_n)
  if (any(min_n == 0)){
    min_n[min_n==0] <- 1
  }
  
  if (direction == "up") {
    mstart <- pmax(1, min_n)
    varvect <- NULL
    while(sum(mstart) < totm){
      nexp <- colSums(mstart * bhg)
      var1 <- totVar(n = nexp)
      varvect <- c(varvect, var1)
      vtemp <- sapply(1:nrow(bhg), FUN = function(x) {
        nexp1 <- nexp + bhg[x, ]
        totVar(n = nexp1) 
      })
      
      d <- var1 - vtemp
      top <- order(-d)
      if (missing(rand)){
        top <- top[mstart[top] < max_n[top]]
        top <- top[steps]
        if (any(d[top] < 0)) warning(paste("step to total: ", sum(mstart),  " increased the minimizing function", sep=""))
      } else {
        top <- top[mstart[top] < max_n[top]][1:rand]
        d[top < 0] <- 0
        p <- d[top]/sum(d[top])
        if(all(is.na(p))) { #what about if more than steps are <0?
          top <- top[steps] #force to add one if all options are a bad solution!#?
          warning(paste("step to total: ", sum(mstart),  " increased the minimizing function", sep=""))
        } else {
          top <- sample(top, size = steps, prob = p)}
      }
      mstart[top] <- mstart[top] + 1
    }
    
  } else if (direction == "down") {
    mstart <- max_n
    while(sum(mstart) > totm){
      nexp <- colSums(mstart * bhg)
      var1 <- totVar(n = nexp)
      vtemp <- sapply(1:nrow(bhg), FUN = function(x) {
        nexp1 <- nexp - bhg[x, ]
        totVar(n = nexp1) 
      })
      d <- var1 - vtemp
      top <- order(-d)
      top <- top[mstart[top] > min_n[top]]
      top <- top[steps]
      mstart[top] <- mstart[top] - 1
    }
  }
  
  names(mstart) <- rownames(bhg)
  nexp <- colSums(mstart * bhg)
  varf <- totVar(n = nexp)
  return(list(alloc = SpecialRound(mstart), totvar = varf))
}



#' Optimal allocation of sample sizes by swapping
#' 
#' This function finds an optimal allocation by swapping one unit at a time
#' between two strata. Beta version: use with caution - needs some extra testing
#' 
#' 
#' @param m Starting values for sample size in each stratum. This should sum to
#' the desired sample size.
#' @param bhg A matrix giving the average number of observations in each
#' elementary group in each stratum. See Zhang 2015 for more details.
#' @param totVar The variance to the index as a function of n (number of
#' observations). See example (?).
#' @param min_n A minimum sample size in each stratum. This can be a single
#' number used in all strata or a vector of numbers (one for each stratum).
#' @param max_n A maximum sample size in each stratum. This can be a single
#' number used in all strata or a vector of numbers (one for each stratum).
#' @param iter Number of unsuccessful iterations for the algorithm to try
#' before stopping. Default is 100.
#' @return Returns a vector with the new sample sizes in each strata
#' 
#' @export
#' 
#' @examples{
#'  
#' # call in test dataset
#' data(priceData)
#' 
#' # Calculate s2 for index
#' s2 <- CalcIndS2(data = priceData, baseVar = "b1", pVar = "p1", groupVar = "varenr", 
#'                 type = "carli")$s2
#' 
#' # Calculate bhg matrix
#' ngh <- table(priceData$nace3, priceData$varenr)
#' mh <- matrix(rep(table(priceData$nace3), 200), 5, 200)
#' bhg <- ngh/mh 
#' 
#' # Specify variance function for n (number of observations) - unweighted
#' VarP <- function(n){
#'   return(sum(s2 / n , na.rm = TRUE))
#' }
#' 
#' # Start values for sample size in each strata with equal numbers (adding to 300) 
#' m <- rep(300/5, 5)
#' 
#' # Run Swap strata
#' a <- SwapStrataInd(m = m, bhg = bhg, totVar = VarP, iter = 100)
#' a
#' }
#' 
SwapStrataInd <- function(m, bhg, totVar, min_n, max_n, iter = 100){
  if (missing(min_n)) min_n <- rep(1, length(m))
  if (missing(max_n)) max_n <- rep(Inf, length(m))
  i <- 1:length(m)
  mstart <- m
  teller <- 0
  teller2 <- 0
  olds2 <- totVar(colSums(mstart * bhg))
  while(teller < iter){
    ss <- sample(i, 2)
    mnew <- mstart
    mnew[ss[1]] <- mnew[ss[1]] + 1
    mnew[ss[2]] <- mnew[ss[2]] - 1
    if (all(mnew >= min_n) & all(mnew <= max_n)) {
      nys2 <- totVar(colSums(mnew * bhg))
      if (olds2 > nys2) {
        mstart <- mnew
        olds2 <- nys2
        teller <- 0
        teller2 <- teller2 + 1
      } else{
        teller <- teller + 1
      }
    }
  }
  cat("There has been ", teller2, " succesful iterations. \n", sep="") 
  return(mstart)
}


#' Internal allocation function
#' 
#' Not sure if used in price index ...
#'
#' @param s2 Sigma squared values
#' @param w Weights vector
#' @param ai Adjustment factor
#' @param totn Total sample size
#'
#' @return ni 
AllocTot <- function(s2, w, ai, totn){
  ws <- w * sqrt(s2) / sqrt(ai)
  ni <- as.vector(totn * ws / sum(ws, na.rm=T))
  return(ni)
}

#' Internal allocation M function
#' 
#' Not sure if actually in use in price index
#'
#' @param bhgData Data giving the average number of price observations in the stratum for group.
#' @param n0 Allocation
#'
#' @return m
AllocM <- function(bhgData, n0){
  bhg <- apply(bhgData, 1:2, mean)
  vh <- rep(NA, nrow(bhg))
  for (i in 1:nrow(bhg)){
    vbh <- stats::cov(t(bhgData[i, , ]))
    vh[i] <- t(rep(1, ncol(bhg)) %*% vbh %*% rep(1, ncol(bhg)))
  }
  invvh <- solve(diag(vh)) # "solve" takes inverse of matrix
  m <- invvh %*% bhg %*% solve((t(bhg) %*% invvh %*% bhg)) %*% n0
  return(m)
}


#' Rounding strata sample sizes
#' 
#' Function for rounding strata sample sizes to ensure the total sample size
#' remains the same
#' 
#' 
#' @param x vector of unrounded strata sample sizes
#' @return vector with round strata sizes where the total sample size is held
#' constant.
#' 
#' @export
#' 
#' @examples{
#'  
#' jj <- c(4.2, 4.2, 4.2, 4.3, 4.3, 4.4, 4.4)
#' SpecialRound(jj)
#' }
SpecialRound <- function (x){
  uk <- x - floor(x)
  difference <- sum(x) - floor(sum(x))
  mynumber <- ifelse(difference < 0.5, ifelse(isTRUE(all.equal(sum(uk), round(sum(uk)))), 
                                              round(sum(uk)), floor(sum(uk))), ceiling(sum(uk)))	
  order.x <- order(-uk)
  x <- x[order(-uk)]
  x <- c(ceiling(x[if(mynumber==0) 0 else seq(mynumber)]), 
         floor(x[(mynumber + 1):length(x)])) 
  return(x[order(order.x)])
}
